[core] make 'PopWorker' to be an async function

SongGuyang opened the pull request:
## Why are these changes needed?
- 'PopWorker' is a synchronous function now. If there are no idle workers, we should call 'PopWorker' repeatedly. And if lease worker with runtime env, we can not catch the failure of runtime env creation and make some other decision, such as rescheduling the task.
- In current PR, we make 'PopWorker' to be an async function. 

Closes https://github.com/ray-project/ray/issues/17154

SongGuyang commented:
@kfstorm @jovany-wang Please help to review this change. I have not modified the worker pool test now. But you can review this PR concurrently with my updates.

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -365,6 +375,13 @@ void WorkerPool::MonitorStartingWorkerProcess(const Process &proc,
+          auto callback = it->second.callback;
+          if (!task_id.IsNil() && callback) {
+            callback(nullptr);
+            it->second.task_id = TaskID::Nil();
+            it->second.callback = PopWorkerCallback();
```
What's the point of resetting these two fields? The entry will be erased anyway.

SongGuyang replied:
If one worker process has multiple workers, it will not be erased.

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -497,13 +514,22 @@ Status WorkerPool::RegisterWorker(const std::shared_ptr<WorkerInterface> &worker
+  bool worker_available = true;
   auto it = state.starting_worker_processes.find(shim_process);
   if (it != state.starting_worker_processes.end()) {
+    auto task_id = it->second.task_id;
+    auto callback = it->second.callback;
```
```suggestion
    const auto &task_id = it->second.task_id;
    const auto &callback = it->second.callback;
```

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -497,13 +514,22 @@ Status WorkerPool::RegisterWorker(const std::shared_ptr<WorkerInterface> &worker
   if (it != state.starting_worker_processes.end()) {
+    auto task_id = it->second.task_id;
+    auto callback = it->second.callback;
+    if (!task_id.IsNil() && callback) {
```
Can we replace it with `if (callback)`?

kfstorm commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -22,6 +22,9 @@ namespace raylet {
 typedef std::tuple<Task, rpc::RequestWorkerLeaseReply *, std::function<void(void)>> Work;
 
+// The information of Work instance which waiting worker popped.
+typedef std::tuple<bool, SchedulingClass, Work> WorkWaitingWorkerPopped;
```
A struct, please.

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -151,19 +151,27 @@ void WorkerPool::SetAgentManager(std::shared_ptr<AgentManager> agent_manager) {
+void WorkerPool::PopWorkerCallbackAsync(const PopWorkerCallback callback,
+                                        std::shared_ptr<WorkerInterface> worker) {
+  if (callback) {
+    io_service_->post([callback, worker]() { callback(worker); });
```
Why do we have to post here?

SongGuyang replied:
I want to make the code execute in order. But I found this bring a lot problem in test case. So I have delete this post.

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -151,19 +151,27 @@ void WorkerPool::SetAgentManager(std::shared_ptr<AgentManager> agent_manager) {
   agent_manager_ = agent_manager;
 }
 
+void WorkerPool::PopWorkerCallbackAsync(const PopWorkerCallback callback,
```
I think this helper method is unnecessary. I saw that some callback invocations do not leverage this method.

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -497,13 +514,22 @@ Status WorkerPool::RegisterWorker(const std::shared_ptr<WorkerInterface> &worker
 }
 
-void WorkerPool::OnWorkerStarted(const std::shared_ptr<WorkerInterface> &worker) {
+bool WorkerPool::OnWorkerStarted(const std::shared_ptr<WorkerInterface> &worker) {
```
I think the meaning of return value is not intuitive. Can we change it to this?
```suggestion
void WorkerPool::OnWorkerStarted(const std::shared_ptr<WorkerInterface> &worker, bool *worker_available) {
```

SongGuyang replied:
I think this way is not convenient. I have add doc to the function definition.

rkooo567 replied:
Why don't we just combine this function with handleWorkerAvailable and make that return true false for worker_available?

rkooo567 replied:
This function is used only once, and `handleWorkerAvailable` is always coming together

SongGuyang replied:
The meaning of `OnWorkerStarted ` return value is that whether the started worker assigned to an async `PopWorker` request or not. So @ericl, does `worker_started_successfully` is a suitable name？Maybe `worker_available ` is better?

SongGuyang replied:
@rkooo567 I don't understand. Do you mean call `OnWorkerStarted ` inside `handleWorkerAvailable `?

rkooo567 replied:
Yeah that's what I meant (and having worker_available as a return value makes more sense in this case)

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -862,21 +889,24 @@ int GetRuntimeEnvHash(const TaskSpecification &task_spec) {
-std::shared_ptr<WorkerInterface> WorkerPool::PopWorker(
-    const TaskSpecification &task_spec) {
+void WorkerPool::PopWorker(const TaskSpecification &task_spec,
+                           const PopWorkerCallback callback) {
```
```suggestion
                           const PopWorkerCallback &callback) {
```

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -497,13 +514,22 @@ Status WorkerPool::RegisterWorker(const std::shared_ptr<WorkerInterface> &worker
+      callback(worker);
+      worker_available = false;
+      it->second.task_id = TaskID::Nil();
+      it->second.callback = PopWorkerCallback();
```
nit
```suggestion
      it->second.callback = nullptr;
```

kfstorm commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -862,21 +889,24 @@ int GetRuntimeEnvHash(const TaskSpecification &task_spec) {
+                                        bool dedicated, const int runtime_env_hash,
+                                        const std::string &serialized_runtime_env,
+                                        const std::string &serialized_runtime_env_context,
+                                        const PopWorkerCallback callback) -> Process {
```
```suggestion
                                        const PopWorkerCallback &callback) -> Process {
```

kfstorm commented on src/ray/raylet/worker_pool.h:
```diff
@@ -431,6 +442,8 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
     std::unordered_set<TaskID> tasks_with_dedicated_workers;
```
Do we still need the concept of "decicated" workers?

SongGuyang replied:
Yes, decicated workers are used for the actor creation tasks with DynamicWorkerOptions. When we pop a idle worker, we need this to distinguish.

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -236,6 +239,11 @@ class ClusterTaskManager : public ClusterTaskManagerInterface {
+  /// Map of lease requests that waiting for workers popped.
+  /// Tasks move from dispatch to this map.
+  /// Tasks can also move from this map to dispatch if workers can not be popped.
+  std::unordered_map<TaskID, WorkWaitingWorkerPopped> tasks_waiting_workers_popped_;
```
```suggestion
  std::unordered_map<TaskID, PendingWorker> pending_workers_;
```

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -236,6 +239,11 @@ class ClusterTaskManager : public ClusterTaskManagerInterface {
 
+  /// Map of lease requests that waiting for workers popped.
+  /// Tasks move from dispatch to this map.
+  /// Tasks can also move from this map to dispatch if workers can not be popped.
```
```suggestion
  /// Tasks can also move from this map to dispatch if workers cannot be popped.
```

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -236,6 +239,11 @@ class ClusterTaskManager : public ClusterTaskManagerInterface {
   std::unordered_map<SchedulingClass, std::deque<Work>> tasks_to_dispatch_;
 
+  /// Map of lease requests that waiting for workers popped.
+  /// Tasks move from dispatch to this map.
```
Do we still need the dispatch queue if we have this queue? It seems that tasks won't spend any time in dispatch (can a pop ever fail now?)

SongGuyang replied:
I have deleted this queue in latest code.

ericl reviewed (commented):
It would be great if we could avoid adding a new queue. I think we could eliminate the dispatch queue if we ensured pops never failed (or you can have an async failure callback handler).

jovany-wang commented on src/ray/raylet/worker_pool.h:
```diff
@@ -350,16 +354,19 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
       std::unordered_map<std::string, std::string> override_environment_variables = {},
-      const std::string &serialized_runtime_env_context = "{}");
+      const std::string &serialized_runtime_env_context = "{}",
+      const PopWorkerCallback callback = PopWorkerCallback());
```
I think `callback` must be non empty. So there's no need to add a default arguement.

jovany-wang commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -308,7 +316,7 @@ class ClusterTaskManager : public ClusterTaskManagerInterface {
       std::shared_ptr<WorkerInterface> worker,
       std::unordered_map<WorkerID, std::shared_ptr<WorkerInterface>> &leased_workers_,
-      std::shared_ptr<TaskResourceInstances> &allocated_instances, const Task &task,
+      std::shared_ptr<TaskResourceInstances> allocated_instances, const Task &task,
```
Hm, why remove the &?  I didn't see the `std::move` when invoking this.

SongGuyang replied:
Revert  to use reference.

SongGuyang commented:
@ericl @kfstorm @jovany-wang This PR has been updated a lot. Please help to review again. @ericl And to avoid adding a new queue, I have added some status to `Work` and reused dispatch queue.

SongGuyang commented:
@simon-mo I remove `blocked_runtime_env_to_skip ` in current PR because `Popworker` is to be a async function. Please help to review.

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -20,7 +20,14 @@ namespace raylet {
+  Task task;
+  rpc::RequestWorkerLeaseReply *reply;
+  std::function<void(void)> callback;
+  bool waiting_worker_popped;
```
```suggestion
  bool waiting_worker_popped = false;
```

stephanie-wang replied:
Could we make this and dispatch_finished into a single enum instead of having two separate flags? This seems less brittle, since we only have one state to check. I don't think it's valid to have waiting_worker_popped && dispatch_finished anyway.

rkooo567 replied:
+1 I think we can just do 

```
enum {
    WAITING # default
    CANCELLED # cancel = true
    SCHEDULED # waiting_worker_popped = true
    DISPATCHED # dispatch_finished = true
}
```

Are there conditions that multiple of them co-exist?

SongGuyang replied:
@rkooo567 Great! Dose the semantics of `SCHEDULED` make confusing？In raylet, `SCHEDULED ` means node selected and resources occupied?

rkooo567 replied:
Isn't the worker popped only when it is scheduled though? Or does the pop mean "it will be eventually be scheduled"?

stephanie-wang replied:
I think we can make it explicit and call it `WAITING_FOR_WORKER` instead of `SCHEDULED`. That seems pretty clear.

SongGuyang replied:
@wuisawesome I found we don't need `SPILLED ` because the item of `Work` will be removed after spilling.

SongGuyang replied:
Just use three status:
```
enum WorkStatus {
  WAITING,
  WAITING_FOR_WORKER,
  CANCELLED,
};
```

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -20,7 +20,14 @@ namespace raylet {
+  rpc::RequestWorkerLeaseReply *reply;
+  std::function<void(void)> callback;
+  bool waiting_worker_popped;
+  bool dispatch_finished;
```
```suggestion
  bool dispatch_finished = false;
```

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -20,7 +20,14 @@ namespace raylet {
+  std::function<void(void)> callback;
+  bool waiting_worker_popped;
+  bool dispatch_finished;
+  bool canceled;
```
```suggestion
  bool canceled = false;
```

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -20,7 +20,14 @@ namespace raylet {
+  bool waiting_worker_popped;
+  bool dispatch_finished;
+  bool canceled;
+};
```
Could we disallow copy and assign on this struct?

SongGuyang replied:
Good for me.

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +246,58 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
-        task_dependency_manager_.RemoveTaskDependencies(
-            task.GetTaskSpecification().TaskId());
+        work->waiting_worker_popped = true;
+        pending_workers_index_.emplace(task_id, work);
```
Could we remove this index and pass in a shared_ptr<Work> to the callback?

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.h:
```diff
@@ -234,7 +236,11 @@ class ClusterTaskManager : public ClusterTaskManagerInterface {
+      tasks_to_dispatch_;
+
+  /// Map of lease requests that waiting for workers popped.
+  std::unordered_map<TaskID, std::shared_ptr<Work>> pending_workers_index_;
```
It seems possible to remove this index.

simon-mo commented:
@SongGuyang as long as 
https://github.com/ray-project/ray/blob/8db61569f11685c816e2cbfb2c01fc45c3164f7c/python/ray/tests/test_runtime_env_complicated.py#L579
is passed you can definitely remove it.

cc @wuisawesome we might want to run scalability test for thsi PR

stephanie-wang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -499,14 +521,22 @@ bool ClusterTaskManager::CancelTask(const TaskID &task_id) {
           task_dependency_manager_.RemoveTaskDependencies(
               task.GetTaskSpecification().TaskId());
         }
+        (*work_it)->canceled = true;
```
Can you add a comment to explain when the task actually gets erased?

SongGuyang replied:
@stephanie-wang I have erased this task in next line? https://github.com/ray-project/ray/blob/7378327c59af9bf48291a17eaee1b43c6ce832d0/src/ray/raylet/scheduling/cluster_task_manager.cc#L540

stephanie-wang replied:
Ah yes, I meant to explain when *work_it gets erased. I think that happens when the dispatch loop gets triggered again, right?

SongGuyang replied:
We have two place to erase task now: `CancelTask` and `erase_from_dispatch_queue_fn`.

stephanie-wang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +246,58 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+              const auto &callback = work->callback;
+              const auto &canceled = work->canceled;
+              if (canceled || !worker) {
+                if (canceled) {
```
If the task was canceled, don't we need to push the worker back to the worker pool somewhere?

SongGuyang replied:
My fault. Let me fix.

stephanie-wang commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -152,8 +152,19 @@ void WorkerPool::SetAgentManager(std::shared_ptr<AgentManager> agent_manager) {
+                                            std::shared_ptr<WorkerInterface> worker,
+                                            Status status) {
+  if (callback) {
+    // Call back this function asynchronously if needed. Such as
```
I don't understand this comment. The code seems to be calling the callback synchronously?

stephanie-wang replied:
This method also seems to be doing very little, could we just inline it wherever it gets called?

stephanie-wang commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -184,9 +198,12 @@ Process WorkerPool::StartWorkerProcess(
+    oss << "Worker not started, " << starting_workers << " workers of language type "
+        << static_cast<int>(language) << " pending registration";
+    RAY_LOG(DEBUG) << oss.str();
+    PopWorkerCallbackExecution(callback, nullptr,
```
Why not wait to call the callback until the starting workers register?

SongGuyang replied:
Yep, when the starting workers number greater than maximum_startup_concurrency_, we will not start an new worker process. We should re-dispatch this task.

SongGuyang replied:
Let me see how to test it.

stephanie-wang replied:
Hmm I guess I thought that now that the callback is async, we should just wait until a worker was available. Right now, the logic will loop:
1. Trigger dispatch/PopWorker.
2. If startup concurrency maxed out, callback with empty worker.
But we could also avoid looping if we did this:
1. Trigger dispatch/PopWorker.
2. If startup concurrency maxed out, wait until a worker is pushed or a worker can be started, then callback.

I don't think it's a good idea to do it in this PR, but it would be good to do it in the future. Can you add as a TODO?

SongGuyang replied:
Sure, Let me add this TODO.

stephanie-wang commented on src/ray/raylet/worker_pool.h:
```diff
@@ -282,13 +288,13 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
-      const TaskSpecification &task_spec,
-      const std::string &allocated_instances_serialized_json = "{}");
+  /// \param callback The callback function that executed when gets the result of
+  /// worker popping.
```
I am a bit confused about the possible triggers for the callback. Can you comment and say what they are? In particular, I'm not sure about what happens when:
1. There's no job config available yet.
2. There's no available worker (there seems to be multiple cases here?)

I might be wrong, but it also doesn't seem like the unit tests in worker_pool_tests check whether we return the right error codes and at the right time in these cases. Can we add tests for these?

SongGuyang replied:
If there's no job config available or the starting workers number greater than maximum_startup_concurrency_, we will callback an empty worker to trigger re-dispatch. And If runtime env creation failed, we should also callback empty worker but trigger re-schedule to other nodes or mark task dead. But I haven't implemented this in current PR, just add some status codes.

stephanie-wang replied:
Thanks, can we make sure to document what the cases are where will trigger the callback? I noticed you added status codes, but it would be good to explain what they mean in the comments.

stephanie-wang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +246,58 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+                // spill this task.
+                ReleaseTaskArgs(task_id);
+                // TODO(guyang.sgy): re-schedule the task by status code.
+                work->waiting_worker_popped = false;
```
Couldn't this hang? Since there is no guarantee DispatchScheduledTasksToWorkers will get called again.

SongGuyang replied:
Good comment! If the status code is JobNotStarted or WorkerPendingRegistration,  maybe `DispatchScheduledTasksToWorkers ` will be triggered by other events? Such as job register or worker register. But If the status code is RuntimeEnvCreationFailed, this will hang. I should do some work here. What do you think?

stephanie-wang replied:
Hmm now that I'm thinking about this more, this might be okay since the previous version of the code also didn't handle the case.

Could you add a `NOTE:` comment to explain that it's assumed that DispatchScheduledTasksToWorkers will get called again?

SongGuyang replied:
I have solved this problem now. If `RuntimeEnvCreationFailed `happens, we will fail the task directly.

stephanie-wang reviewed (approved):
Mainly some questions about corner cases.

rkooo567 commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -152,8 +152,19 @@ void WorkerPool::SetAgentManager(std::shared_ptr<AgentManager> agent_manager) {
+  if (callback) {
+    // Call back this function asynchronously if needed. Such as
+    // io_service_->post([callback, worker, status]() { callback(worker, status); });
+    callback(worker, status);
```
I think we should always post it instead of calling in the same stack. This can cause many unexpected weird issues

SongGuyang replied:
I agree that using post is more reasonable. But I found both `worker_pool_test` and `cluster_task_manager_test` are synchronous test. I should refactor tests any more.

rkooo567 replied:
I remember there were cases like https://github.com/ray-project/ray/issues/2959. I guess it is not happening in the current code structure, but there's nothing to lose by being more careful

rkooo567 commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -372,8 +390,14 @@ void WorkerPool::MonitorStartingWorkerProcess(const Process &proc,
+              << ") have not registered to raylet within timeout.";
+          RAY_LOG(INFO) << oss.str();
+          const auto &callback = it->second.callback;
+          if (callback) {
```
Use `PopWorkerCallbackExecution `

rkooo567 commented on src/ray/common/status.cc:
```diff
@@ -102,6 +106,9 @@ std::string Status::CodeAsString() const {
       {StatusCode::TransientObjectStoreFull, STATUS_CODE_TRANSIENT_OBJECT_STORE_FULL},
+      {StatusCode::JobNotStarted, STATUS_CODE_JOB_NOT_STARTED},
+      {StatusCode::WorkerPendingRegistration, STATUS_CODE_WORKER_PENDING_REGISTRATION},
+      {StatusCode::RuntimeEnvCreationFailed, STATUS_CODE_RUNTIME_ENV_CREATION_FAILED},
```
It looks like we don't catch all these status inside a caller and do special action. Why don't we just use one status and write better error message inside? (e.g., WorkerPoolStartupNotReady Status?)

SongGuyang replied:
Code updated. I have added some different branches by status.

rkooo567 commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +246,58 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+                // No worker available, we won't be able to schedule any kind of task.
+                // Worker processes spin up pretty quickly, so it's not worth trying to
+                // spill this task.
+                ReleaseTaskArgs(task_id);
```
It looks like
```cpp
      if (!spec.GetDependencies().empty()) {
        task_dependency_manager_.RemoveTaskDependencies(
            task.GetTaskSpecification().TaskId());
```
is always called even when worker == nullptr. Don't we need to also call it in this block?

SongGuyang replied:
There is a if to process worker == nullptr?
https://github.com/ray-project/ray/blob/7378327c59af9bf48291a17eaee1b43c6ce832d0/src/ray/raylet/scheduling/cluster_task_manager.cc#L261

rkooo567 replied:
Oh I meant this code block is called always in the original code (regardless of worker == nullptr or worker != nullptr)

SongGuyang replied:
Great! My bug. Thank you. I fixed this.

wuisawesome reviewed (changes requested):
I really like the idea of using a callback for for the worker pool, but I think there are lots of edge cases to handle here, and we should have a lot more test coverage for this.

I pointed out a few specific places, but in general, can we also:
1. In the cluster task manager tests, assert that we're only attempting to send 1 reply?
2. In the worker pool, assert that we always call the callback immediately within PopWorker, _except_ when we actually start/register a worker process (in which case we should assert the callback isn't called immediately) where we should assert it's only called after registration is complete?

wuisawesome commented:
Can we also run the [distributed benchmarks](https://github.com/ray-project/ray/blob/master/benchmarks/distributed/test_distributed.py) (or something similar) to show that this doesn't cause issues at large scale? We've been running into lots of problems around scale in this area lately.

SongGuyang commented:
Thanks for all the comments that are very useful to me. Let me take some time to address all the comments. I agree that we should do more tests for this PR to make sure we don't cause new issues.

SongGuyang commented:
@edoakes I removed `runtime_env_setup_failed_callback` in current PR. Please help to review.

SongGuyang commented:
@wuisawesome Can you tell me how to run  distributed benchmarks? I need to find 65 machines to run this?

SongGuyang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -237,6 +232,11 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+          task_dependency_manager_.RemoveTaskDependencies(
+              task.GetTaskSpecification().TaskId());
+        }
+        work_it = dispatch_queue.erase(work_it);
```
Code outdated.

SongGuyang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -237,6 +232,11 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+              task.GetTaskSpecification().TaskId());
+        }
+        work_it = dispatch_queue.erase(work_it);
       } else {
```
code outdated.

SongGuyang commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -184,9 +198,12 @@ Process WorkerPool::StartWorkerProcess(
+        << static_cast<int>(language) << " pending registration";
+    RAY_LOG(DEBUG) << oss.str();
+    PopWorkerCallbackExecution(callback, nullptr,
+                               Status::WorkerPendingRegistration(oss.str()));
```
Yes, I have renamed this status to `SoManyStartingWorkerProcesses`. Do you think `WorkerStartupRateLimit ` is a better name?

SongGuyang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -237,6 +232,11 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
         }
+        if (!spec.GetDependencies().empty()) {
+          task_dependency_manager_.RemoveTaskDependencies(
+              task.GetTaskSpecification().TaskId());
```
I just copied previous code.

SongGuyang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +246,58 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+        worker_pool_.PopWorker(
+            spec,
+            [this, allocated_instances, task_id, runtime_env_worker_key](
+                const std::shared_ptr<WorkerInterface> worker, Status status) {
```
I don't understand. In my latest code, I bind task to an worker process. Not receive multiple worker processes?

stephanie-wang commented on src/ray/common/status.h:
```diff
@@ -101,6 +101,12 @@ enum class StatusCode : char {
   ObjectStoreFull = 24,
   TransientObjectStoreFull = 25,
+  // worker pool status
+  JobNotStarted = 30,
```
```suggestion
  JobConfigMissing = 30,
```

stephanie-wang commented on src/ray/common/status.h:
```diff
@@ -101,6 +101,12 @@ enum class StatusCode : char {
+  // worker pool status
+  JobNotStarted = 30,
+  OwnerFailed = 31,
+  SoManyStartingWorkerProcesses = 32,
```
```suggestion
  TooManyStartingWorkerProcesses = 32,
```

stephanie-wang commented on src/ray/common/status.h:
```diff
@@ -101,6 +101,12 @@ enum class StatusCode : char {
   ObjectAlreadySealed = 23,
   ObjectStoreFull = 24,
   TransientObjectStoreFull = 25,
+  // worker pool status
```
Hey, can we actually not put these as part of the common StatusCodes? Let's make these an enum member of the WorkerPool class instead. These error codes are pretty specific to the worker pool logic. We can change the callback signature to return this enum instead of a Status.

In general, we should reserve the common StatusCodes for cross-process communication only.

SongGuyang replied:
Good point. I added an enum `PopWorkerStatus`.

SongGuyang commented:
@stephanie-wang @wuisawesome @ericl @rkooo567 Code has been updated. Please help to review again. Thanks.

stephanie-wang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +239,81 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+              bool canceled = work->status == WorkStatus::CANCELLED;
+              const auto &task = work->task;
+              const auto &spec = task.GetTaskSpecification();
+              bool used = false;
```
```suggestion
              bool dispatched = false;
```

stephanie-wang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +239,81 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+                cluster_resource_scheduler_->ReleaseWorkerResources(allocated_instances);
+                // Release pinned task args.
+                ReleaseTaskArgs(task_id);
+                if (status.IsOwnerFailed() && !canceled) {
+                  erase_from_dispatch_queue_fn(work, scheduling_class);
+
+                } else if (status.IsRuntimeEnvCreationFailed()) {
+                  CancelTask(task_id, true);
+                } else {
+                  work->status = WorkStatus::WAITING;
+                }
```
Can you add some comments here to explain why the different cases are handled this way?

stephanie-wang replied:
Also, is this behavior the same as the previous code? I didn't think that we canceled tasks if the runtime env creation failed, but not sure.

SongGuyang replied:
Sure, I added some comments.

SongGuyang replied:
The runtime env failed behavior has been merged in this PR. https://github.com/ray-project/ray/commit/7007c6271d4becb762ab33d65a6841824a15134e

stephanie-wang commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -153,10 +153,19 @@ void WorkerPool::SetAgentManager(std::shared_ptr<AgentManager> agent_manager) {
+                                               Status status) {
+  if (callback) {
+    // Call back this function asynchronously to make sure executed in different stack.
+    io_service_->post([callback, worker, status]() { callback(worker, status); });
```
Don't we need to check the return value of the callback? Otherwise this seems like we could leak a worker (ClusterTaskManager isn't using the worker, but the worker is still marked as occupied).

stephanie-wang replied:
Can we also make sure to add a unit test for this case?

SongGuyang replied:
Great！My bug. Let me fix.

stephanie-wang commented on src/ray/raylet/worker_pool.h:
```diff
@@ -40,6 +41,9 @@ namespace raylet {
 using WorkerCommandMap =
     std::unordered_map<Language, std::vector<std::string>, std::hash<int>>;
 
+using PopWorkerCallback =
```
Please add a comment explaining the return value.

stephanie-wang commented on src/ray/raylet/worker_pool.h:
```diff
@@ -528,6 +550,15 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
+                                     std::shared_ptr<WorkerInterface> worker,
+                                     Status status = Status::OK());
+
+  void TryToCallbackTask(
```
```suggestion
  void TryToAssignTaskToDedicatedWorker(
```

Please also add a comment to explain what this function does.

stephanie-wang commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -246,38 +239,81 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+            spec,
+            [this, allocated_instances, task_id, scheduling_class, work](
+                const std::shared_ptr<WorkerInterface> worker, Status status) -> bool {
+              const auto &reply = work->reply;
```
Suggest making this lambda a method of the ClusterTaskManager class (just easier to read).

SongGuyang replied:
I have a lot variables captured. And lambda is used a lot in our project. Maybe just leave it a lambda ok?

stephanie-wang commented:
Is it really necessary to run the distributed tests, @wuisawesome ? I think it's okay to merge without that. The nightly tests should be enough, right?

wuisawesome commented:
@stephanie-wang the full scale distributed one (which is the most likely one to break from a change like this), only runs twice a week. 

If you want to merge without running it, I don't feel too strongly about it, but this is the type of commit we'll likely have to revert.

stephanie-wang commented:
> @stephanie-wang the full scale distributed one (which is the most likely one to break from a change like this), only runs twice a week.
> 
> If you want to merge without running it, I don't feel too strongly about it, but this is the type of commit we'll likely have to revert.

Can we think of a Python and/or C++ test that we should add instead? It just seems strange to me that we would need to rerun large-scale tests for such a localized change. If there have been issues with WorkerPool, then it's best for long-term developer productivity if we can figure out a way to quickly and safely merge changes.

wuisawesome commented:
> Can we think of a Python and/or C++ test that we should add instead? It just seems strange to me that we would need to rerun large-scale tests for such a localized change. 

I 100% agree we should have these tests, but until we have a way writing these tests, the reality is that the large scale tests are the only ones catching these bugs. 

I think what we ideally want is a test that can flood the cluster task manager with tasks, and ensure that the correct number of worker processes are started, and things are dispatched accordingly. That being said, I'm not sure we'll be able to write that test without restructuring the worker pool, which isn't a priority. If there's an easy way to write that type of test we should totally do it.

rkooo567 commented:
@wuisawesome It sounds like we can totally write that in python tests?

stephanie-wang commented:
It seems like we could do that with a Python test if we just collect and expose metrics about how many workers are started. Would that work?

rkooo567 commented:
I think as a short term workaround, we can write them to debug_state.txt and read it within tests. (We don’t have an easy way to access metrics programatically rn)

wuisawesome commented:
Yeah if we can write a python test to flood the raylet that would be great. It's a little hard to do though, since i believe both the core worker and gcs try to avoid flooding the raylet. 

I think the test we'd want to write is: give a raylet 4 cpus, then send it 10,000 lease requests for 0.25 CPUs each. It should only start 16 workers. Unfortunately, on a cluster with 4 cpus, I believe GCS actor creation will only send 16 lease requests to the raylet, and task based workload from a driver would only send 17.

stephanie-wang commented:
> Yeah if we can write a python test to flood the raylet that would be great. It's a little hard to do though, since i believe both the core worker and gcs try to avoid flooding the raylet.
> 
> I think the test we'd want to write is: give a raylet 4 cpus, then send it 10,000 lease requests for 0.25 CPUs each. It should only start 16 workers. Unfortunately, on a cluster with 4 cpus, I believe GCS actor creation will only send 16 lease requests to the raylet, and task based workload from a driver would only send 17.

I think we can trigger this behavior if we assign each lease request a random CPU requirement between say [0.25, 0.26]. I believe that will force one lease request/CPU value.

wuisawesome commented:
Yeah good point. I'm in favor of trying it out. (We can also test it against 54d66ac6378ece17d609fae349230ca1bb341742 which is known to suffer from the worker pool issues).

SongGuyang commented:
@wuisawesome @stephanie-wang So, do I need to do this benchmark test before current PR merged？

SongGuyang commented:
@stephanie-wang @wuisawesome Can you help to take a look at this PR again? I have addressed your comments and added some tests. Please help to merge this PR as soon as possible. I have meet a lot conflicts during this two weeks. Thank you.

stephanie-wang commented on src/ray/raylet/worker_pool.h:
```diff
@@ -41,8 +41,28 @@ namespace raylet {
+  // Worker process startup rate is limited.
+  // A nullptr worker will be returned with callback.
+  TooManyStartingWorkerProcesses = 2,
+  // Worker process has been started, but the worker registers back to raylet timeout.
```
```suggestion
  // Worker process has been started, but the worker did not register at the raylet within the timeout.
```

stephanie-wang commented on src/ray/raylet/worker_pool_test.cc:
```diff
@@ -1324,66 +1319,119 @@ TEST_F(WorkerPoolTest, StartWorkWithDifferentShimPid) {
-
-  // Pop worker for a task with runtime env.
-  auto popped_worker = worker_pool_->PopWorker(normal_task_spec);
+TEST_F(WorkerPoolTest, WorkerNoLeaks) {
```
Thanks for adding this test!

raulchen commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -695,19 +726,49 @@ void WorkerPool::PopDeleteWorker(
+void WorkerPool::TryToAssignTaskToDedicatedWorker(
+    std::unordered_map<Process, TaskWaitingForWorkerInfo> &starting_workers_to_tasks,
+    const Process &proc, const std::shared_ptr<WorkerInterface> &worker,
+    const PopWorkerStatus &status, bool *found, bool *used, TaskID *task_id) {
```
`used` is called `dispatched` in the header file.

raulchen commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -262,38 +255,128 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+        auto &owner_address = spec.CallerAddress();
+        worker_pool_.PopWorker(
+            spec,
+            [this, task_id, scheduling_class, work, is_detached_actor, owner_address](
```
I'd suggest making this callback a class method as well, as this callback is getting quiet complex.
Imagine next time when someone changes the captured arguments, they may have to change the whole function due to indent changes.

raulchen commented on src/ray/raylet/worker_pool.h:
```diff
@@ -40,13 +41,30 @@ namespace raylet {
 };
 
+/// \Return True if any task dispatched successfully to the worker. Otherwise, return
+/// false.
```
`/// \Return true if the worker was used. Otherwise, return false and the worker will be returned to the worker pool.`
Because from the worker pool's perspective, it should only care the worker, not the task.

raulchen commented on src/ray/raylet/worker_pool.h:
```diff
@@ -290,13 +320,13 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
+  /// \param callback The callback function that executed when gets the result of
+  /// worker popping.
+  /// \param allocated_instances_serialized_json The allocated resouce
+  /// instances json string. \return An idle worker with the requested task spec. Returns
```
`\return Void. `
Also this comment has diverged from the interface one. I'd suggest just putting "See interface.` here.

SongGuyang replied:
Good point.

rkooo567 reviewed (changes requested):
I don't see the python tests that handle https://github.com/ray-project/ray/pull/17202#issuecomment-889556278

Is it possible to handle this? Also, @wuisawesome can you guide him what needs to be tested? Also, check https://github.com/ray-project/ray/pull/17422 to see how to test # workers started.

raulchen commented:
> I don't see the python tests that handle #17202 (comment)

Is it possible to handle this? Also, @wuisawesome can you guide him what needs to be tested? Also, check #17422 to see how to test # workers started.

@rkooo567 I don't understand the purpose of that test. Do you want to make sure that the number of started workers are no more than 16?

raulchen commented on src/ray/raylet/worker_pool.h:
```diff
@@ -535,6 +569,26 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
+  /// `PopWorkerCallback`. \param found  Whether the related task found or not. \param
+  /// dispatched Whether the related task dispatched or not. \param task_id  The related
+  /// task id.
+  void TryToAssignTaskToDedicatedWorker(
```
I think something like `FindWorkerAndInvokePopWorkerCallbackForProcess` will be clearer.

raulchen commented on src/ray/raylet/worker_pool.h:
```diff
@@ -535,6 +569,26 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
+  /// the task. \param status The pop worker status which will be forwarded to
+  /// `PopWorkerCallback`. \param found  Whether the related task found or not. \param
+  /// dispatched Whether the related task dispatched or not. \param task_id  The related
+  /// task id.
```
better rename `dispatched` to `worker_used`: "Whether the worker is used by the task, only valid when found is true."

rkooo567 commented:
> @rkooo567 I don't understand the purpose of that test. Do you want to make sure that the number of started workers are no more than 16?

@wuisawesome can explain this better, but in our stress tests, we've seen worker pools start too many workers in some situations. Since this PR changes the worker pool behavior a lot, we'd like to make sure it won't introduce the new bugs from that workload (that's basically why we were asking to run some release tests).

wuisawesome commented:
Yeah I don't have a complete understanding of what the previous bug was, because we just bisected and removed the offending commit, but what we saw was that `PopWorker` was getting called more often and it resulted in thousands of worker processes starting on a single node and not being able to register.

rkooo567 commented:
@wuisawesome do you have a specific test scenario in mind?

SongGuyang commented:
@wuisawesome @rkooo567 @raulchen Test has been added https://github.com/ray-project/ray/pull/17202/commits/fd9016d6e361c80f18002eaaa2268c8d108af610#diff-b61a4cb137dafbbffcd3ae212f8d6ddde2b3482da1e1e09d22d67e53a03ea0b8.

SongGuyang commented:
I'm so depressed that conflicting files appears again. I have fixed conflicting so many times.

rkooo567 commented:
@SongGuyang Sorry about that :(. But it seems to be almost there to go!!

rkooo567 commented:
I am down to merge it after @wuisawesome has a final go. But please let me merge the PR so that we can make sure the nightly tests are reliable after this PR is merged.

raulchen commented:
@rkooo567 are you going to manually run a nightly test against this PR?

rkooo567 commented:
I will just merge it on the day we don’t merge a big core changes. (We have other huge core change that could be merged soon, so I’d like to make sure they have enough gaps before merged)

raulchen commented:
@rkooo567 so are you going to merge this PR tomorrow? We don't want to hold this PR for too long, as it keeps getting conflicts.

rkooo567 commented:
Yeah if test result is good and Alex gives the final approval, I will merge it tmrw

rkooo567 commented:
Btw, sorry for the delay in merging it. We've had bad incidents on this worker pool layer recently, so we were being more careful in reviewing the PR this time. Also, I am not sure if it was possible in this case, but it'd be great if we could split the PR into multiple of them because if the PR contains 1000+ lines of code, it becomes very hard to provide quality code review (this is something our end also should improve), which means it can take even longer time to merge. Based on my experience if the PR has more than 1000 lines of code change, it'd expect to take normally 2~3 weeks unless it is just refactoring.

raulchen commented:
@rkooo567 Understood. Originally, we thought this was going to be a small change (just making a function async). But it turned out that a lot of related code (cluster_task_manager for example) needed to be changed as well. And task dispatching and worker managing logic became slightly different. Also, another reason why it's huge is that a lot of sync testing code needed to be async now.

SongGuyang commented:
The core changes only about 300 lines. Most of the change lines are tests code and new tests. And I can't split this PR because I just do one thing: make `PopWorker` to be async. No other irrelevant code.

SongGuyang commented:
All checks have passed.

rkooo567 commented:
Plan to merge at 5pm today.

rkooo567 commented:
Thanks for the PR again @SongGuyang! I will have a close look at the nightly test and lyk how this goes!

SongGuyang commented:
@rkooo567 Thank you!

jovany-wang commented on src/ray/raylet/worker_pool_test.cc:
```diff
@@ -1244,78 +1311,127 @@ TEST_F(WorkerPoolTest, StartWorkWithDifferentShimPid) {
       ActorID::Nil(), Language::JAVA, JOB_ID, actor_id, actor_jvm_options, task_id);
-  ASSERT_EQ(worker_pool_->PopWorker(java_task_spec), nullptr);
+  worker = worker_pool_->PopWorkerSync(java_task_spec);
+  ASSERT_NE(worker, nullptr);
```
I believe this change didn't make sense, because this test is aiming to test shim pid mechanism.

Since we integrated startup token mechanism, this case should be removed.
